# Homework No. 5

The following jupyter notebook contains the results of the fifth homework for the course: "Data Visualization" and was created by Franz Just.

To execute the jupyter notebook, the following libraries are required to install:
- pandas
- seaborn
- matplotlib
- numpy

The dataset that is used was published by **Jack Daoud** on **Kaggle** and can be found through the following **URL:** https://www.kaggle.com/jackdaoud/marketing-data.

In [3]:
%matplotlib inline
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import calendar
import numpy as np
import warnings
import networkx as nx
warnings.filterwarnings('ignore')

In [4]:
df = pd.read_csv("dataset.csv", index_col = None)
df = df.rename(columns={" Income ": "Income"})
df["Income"] = df["Income"].replace("[$,]", "", regex=True).astype(float)
df = df.dropna()
ID = df["ID"].tolist()
df1 = df[["Income", "Kidhome", "Teenhome", "Year_Birth", "Recency", "NumDealsPurchases", "NumCatalogPurchases", "NumStorePurchases", "NumWebPurchases", "NumWebVisitsMonth"]]
df_1 = df1.values.tolist()

In order to calculate the corresponding distances between several customers, a sub-dataset was created with the following columns:

- Income
- Kidhome
- Teenhome
- Year_Birth
- Recency
- NumDealsPurchases
- NumCatalogPurchases
- NumStorePurchases
- NumWebPurchases
- NumWebVisitsMonth

## 1. Manhattan Distance 

In the following, the manhattan distance between the customers will be calculated. The results are hereby visualized in a distance matrix, where the diagonal corresponds to the distance between the same customers and is therefore 0.

In [5]:
from scipy.spatial import distance_matrix
# In terms of the performance, not all rows and columns are visualized. 
# To visualize the whole table, the following two lines have to be uncommented
# pd.set_option('display.max_rows', None)
# pd.set_option('display.max_columns', None)
md = distance_matrix(df_1, df_1, p=1)
df_md = pd.DataFrame(md, columns = ID)
df_md.index = ID
df_md.head()

1826     1        10476    1386     5371     7348     4073     1991   \
1826       0.0  27762.0  17586.0  52383.0  63396.0  13161.0  21305.0  39920.0   
1      27762.0      0.0  10190.0  24641.0  35660.0  14615.0   6491.0  12178.0   
10476  17586.0  10190.0      0.0  34815.0  45836.0   4431.0   3729.0  22352.0   
1386   52383.0  24641.0  34815.0      0.0  11027.0  39246.0  31130.0  12463.0   
5371   63396.0  35660.0  45836.0  11027.0      0.0  50265.0  42149.0  23486.0   

         4047     9477   ...    3363     8595     7232     7829     9977   \
1826   19540.0  19540.0  ...  64825.0  42523.0  42523.0  48378.0   6049.0   
1       8248.0   8248.0  ...  37089.0  14787.0  14787.0  20624.0  21929.0   
10476   1958.0   1958.0  ...  47265.0  24961.0  24961.0  30794.0  11759.0   
1386   32883.0  32883.0  ...  12454.0  10066.0  10066.0   4337.0  46552.0   
5371   43902.0  43902.0  ...   1465.0  21077.0  21077.0  15358.0  57559.0   

         10142    5263     22       528      4070   
1826   18477.0  53903.0  38645.0  19130.0  10151.0  
1       9509.0  26167.0  10905.0   8852.0  37897.0  
10476    919.0  36343.0  21087.0   1578.0  27729.0  
1386   34130.0   1530.0  13956.0  33477.0  62514.0  
5371   45131.0   9699.0  24955.0  44474.0  73531.0  

[5 rows x 2216 columns]

To further analyze the data, sorting can for instance be conducted. That means, If we want to, for example, find the most similar customers (according to our data) to the customer with the ID: 1826, we can sort our data.

In [6]:
df_md = df_md.sort_values(by =[1826])
df_md.head()

1826     1        10476    1386     5371     7348     4073     1991   \
1826    0.0  27762.0  17586.0  52383.0  63396.0  13161.0  21305.0  39920.0   
3725   48.0  27784.0  17612.0  52415.0  63436.0  13189.0  21327.0  39952.0   
2088  181.0  27937.0  17765.0  52562.0  63571.0  13336.0  21476.0  40099.0   
9757  181.0  27937.0  17765.0  52562.0  63571.0  13336.0  21476.0  40099.0   
6906  218.0  27960.0  17774.0  52587.0  63606.0  13345.0  21477.0  40124.0   

        4047     9477   ...    3363     8595     7232     7829    9977   \
1826  19540.0  19540.0  ...  64825.0  42523.0  42523.0  48378.0  6049.0   
3725  19558.0  19558.0  ...  64861.0  42559.0  42559.0  48396.0  6089.0   
2088  19711.0  19711.0  ...  64804.0  42502.0  42502.0  48361.0  6026.0   
9757  19711.0  19711.0  ...  64804.0  42502.0  42502.0  48361.0  6026.0   
6906  19724.0  19724.0  ...  64889.0  42587.0  42587.0  48408.0  6117.0   

        10142    5263     22       528      4070   
1826  18477.0  53903.0  38645.0  19130.0  10151.0  
3725  18511.0  53939.0  38685.0  19166.0  10129.0  
2088  18444.0  53882.0  38622.0  19099.0   9988.0  
9757  18444.0  53882.0  38622.0  19099.0   9988.0  
6906  18547.0  53967.0  38713.0  19200.0   9977.0  

[5 rows x 2216 columns]

By sorting the column, we can see that for instance the customer with the ID 3725 is pretty close to the customer with the ID 1826. 

## 2. Euclidean Distance

In the following, the euclidean distance between the customers will be calculated. The results are hereby visualized in a distance matrix again, where the diagonal corresponds to the distance between the same customers and is therefore 0.

In [7]:
dd = distance_matrix(df_1, df_1, p=2)
df_ed = pd.DataFrame(dd, columns = ID)
df_ed.index = ID
df_ed.head()

1826          1             10476         1386          5371   \
1826       0.000000  27744.001946  17568.004326  52361.000840  63361.003354   
1      27744.001946      0.000000  10176.001965  24617.002275  35617.011722   
10476  17568.004326  10176.001965      0.000000  34793.001782  45793.010908   
1386   52361.000840  24617.002275  34793.001782      0.000000  11000.022318   
5371   63361.003354  35617.011722  45793.010908  11000.022318      0.000000   

              7348          4073          1991          4047          9477   \
1826   13144.005896  21271.008321  39904.000601  19511.007355  19511.007355   
1      14600.001610   6473.008342  12160.003372   8233.003705   8233.003705   
10476   4424.003052   3703.020254  22336.002149   1943.012609   1943.012609   
1386   39217.002231  31090.006079  12457.000321  32850.003973  32850.003973   
5371   50217.010295  42090.016429  23457.010509  43850.014778  43850.014778   

       ...         3363          8595          7232          7829   \
1826   ...  64705.076501  42406.116198  42406.116198  48195.152993   
1      ...  36961.135819  14662.341116  14662.341116  20451.331937   
10476  ...  47137.107198  24838.202209  24838.202209  30627.215283   
1386   ...  12344.399135   9955.494463   9955.494463   4167.715561   
5371   ...   1347.728088  20955.240228  20955.240228  15166.584454   

              9977          10142         5263          22            528    \
1826    5934.829989  18359.269021  53779.092359  38525.128514  19016.259937   
1      21810.228312   9385.535414  26035.194507  10781.465902   8728.579037   
10476  11634.432818    797.402659  36211.141007  20957.242734   1451.530227   
1386   46427.106974  34002.146962   1421.487953  13836.358480  33345.150487   
5371   57427.088094  45002.111850   9582.519345  24836.201098  44345.112944   

              4070   
1826   10036.491369  
1      37780.130849  
10476  27604.180825  
1386   62397.079210  
5371   73397.069839  

[5 rows x 2216 columns]

## 3. Cosine Similarity

In the following, the cosine similarity between the customers will be calculated. The results are hereby visualized in a similarity matrix, where the diagonal corresponds to the similarity between the same customers and is therefore 1.

In [8]:
from numpy import dot
from numpy.linalg import norm
X_CS = df_1

arr_cs = np.zeros((len(X_CS), len(X_CS)))
for i in range(len(X_CS)):
    for j in range(len(X_CS)):
        arr_cs[(i),(j)] = dot(X_CS[i], X_CS[j])/(norm(X_CS[i])*norm(X_CS[j]))

In [9]:
df_cs = pd.DataFrame(arr_cs, columns = ID)
df_cs.index = ID
df_cs.head()

1826      1         10476     1386      5371      7348      4073   \
1826   1.000000  0.999938  0.999983  0.999305  0.997611  0.999992  0.999972   
1      0.999938  1.000000  0.999986  0.999658  0.998317  0.999975  0.999993   
10476  0.999983  0.999986  1.000000  0.999507  0.998000  0.999998  0.999999   
1386   0.999305  0.999658  0.999507  1.000000  0.999492  0.999449  0.999557   
5371   0.997611  0.998317  0.998000  0.999492  1.000000  0.997885  0.998101   

          1991      4047      9477   ...     3363      8595      7232   \
1826   0.999789  0.999978  0.999978  ...  0.997212  0.999727  0.999727   
1      0.999956  0.999990  0.999990  ...  0.997978  0.999924  0.999924   
10476  0.999893  1.000000  1.000000  ...  0.997632  0.999846  0.999846   
1386   0.999860  0.999532  0.999532  ...  0.999294  0.999899  0.999899   
5371   0.998819  0.998050  0.998050  ...  0.999973  0.998944  0.998944   

          7829      9977      10142     5263      22        528       4070   
1826   0.999588  0.999998  0.999978  0.999181  0.999809  0.999976  0.999996  
1      0.999844  0.999956  0.999988  0.999568  0.999963  0.999990  0.999907  
10476  0.999738  0.999991  0.999999  0.999401  0.999906  0.999998  0.999965  
1386   0.999959  0.999369  0.999525  0.999990  0.999838  0.999535  0.999210  
5371   0.999174  0.997731  0.998037  0.999581  0.998762  0.998058  0.997437  

[5 rows x 2216 columns]

# NEW DATASET

In order to calculate the simple matching coefficient and the jaccard similarity, an updatet version of the previous dataset was used. 

In [10]:
df_2 = pd.read_csv("dataset_second.csv", index_col = None)
df_2 = df_2.dropna()
df_2.head()

Income  Kidhome  Teenhome  Recency  MntWines  MntFruits  MntMeatProducts  \
0  58138.0        0         0       58       635         88              546   
1  46344.0        1         1       38        11          1                6   
2  71613.0        0         0       26       426         49              127   
3  26646.0        1         0       26        11          4               20   
4  58293.0        1         0       94       173         43              118   

   MntFishProducts  MntSweetProducts  MntGoldProds  ...  marital_Together  \
0              172                88            88  ...                 0   
1                2                 1             6  ...                 0   
2              111                21            42  ...                 1   
3               10                 3             5  ...                 1   
4               46                27            15  ...                 0   

   marital_Widow  education_2n Cycle  education_Basic  education_Graduation  \
0              0                   0                0                     1   
1              0                   0                0                     1   
2              0                   0                0                     1   
3              0                   0                0                     1   
4              0                   0                0                     0   

   education_Master  education_PhD  MntTotal  MntRegularProds  \
0                 0              0      1529             1441   
1                 0              0        21               15   
2                 0              0       734              692   
3                 0              0        48               43   
4                 0              1       407              392   

   AcceptedCmpOverall  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  

[5 rows x 39 columns]

In [18]:
df_sm = df_2[["Response", "marital_Married", "marital_Single", "marital_Together", "education_Graduation", "education_Master", "education_PhD"]]
df_sm = df_sm.values.tolist()

For the calcualtions, the following columns with their binary values were hereby considered:
- Response
- marital_Married
- marital_Single
- marital_Together
- education_Graduation
- education_Master
- education_PhD

## 4. Simple Matching

In the following, the simple matching coefficient between the customers will be calculated. The results are hereby visualized in a similarity matrix, where the diagonal corresponds to the similarity between the same customers and is therefore 1.

In [12]:
def simple_matching(x,y):
    x_test = np.logical_not(x).tolist()
    x_n = [int(val) for val in x_test]
    y_test = np.logical_not(y).tolist()
    y_n = [int(val) for val in y_test]
    its = np.logical_and(x, y)
    its_n = np.logical_and(x_n, y_n)
    union = np.logical_or(x, y)
    union_n = np.logical_or(x_n, y_n)
    sim = (its.sum() + its_n.sum()) / (float(union.sum()) + its_n.sum())
    return sim

In [13]:
X_SM = df_sm

arr_sm = np.zeros((len(X_SM), len(X_SM)))
for i in range(len(X_SM)):
    for j in range(len(X_SM)):
        arr_sm[(i),(j)] = simple_matching(X_SM[i], X_SM[j])

In [14]:
df_smc = pd.DataFrame(arr_sm)
df_smc.head()

0         1         2         3         4         5         6     \
0  1.000000  0.857143  0.571429  0.571429  0.285714  0.285714  0.714286   
1  0.857143  1.000000  0.714286  0.714286  0.428571  0.428571  0.857143   
2  0.571429  0.714286  1.000000  1.000000  0.428571  0.714286  0.857143   
3  0.571429  0.714286  1.000000  1.000000  0.428571  0.714286  0.857143   
4  0.285714  0.428571  0.428571  0.428571  1.000000  0.428571  0.571429   

       7         8         9     ...      2195      2196      2197      2198  \
0  0.285714  0.428571  0.285714  ...  0.571429  0.857143  0.571429  0.857143   
1  0.428571  0.285714  0.428571  ...  0.714286  1.000000  0.714286  1.000000   
2  0.428571  0.571429  0.714286  ...  0.714286  0.714286  0.428571  0.714286   
3  0.428571  0.571429  0.714286  ...  0.714286  0.714286  0.428571  0.714286   
4  1.000000  0.571429  0.714286  ...  0.714286  0.428571  0.428571  0.428571   

       2199      2200      2201      2202      2203      2204  
0  0.571429  0.571429  0.285714  0.714286  0.285714  0.428571  
1  0.714286  0.714286  0.428571  0.857143  0.428571  0.285714  
2  0.714286  0.714286  0.714286  0.857143  0.714286  0.285714  
3  0.714286  0.714286  0.714286  0.857143  0.714286  0.285714  
4  0.714286  0.714286  0.714286  0.571429  0.428571  0.857143  

[5 rows x 2205 columns]

## 5. Jaccard Similarity

In the following, the Jaccard Similarity between the customers will be calculated. The results are hereby visualized in a similarity matrix, where the diagonal corresponds to the similarity between the same customers and is therefore 1.

In [19]:
X_JS = df_sm

def jaccard_binary(x,y):
    its = np.logical_and(x, y)
    union = np.logical_or(x, y)
    sim = its.sum() / float(union.sum())
    return sim

arr_js = np.zeros((len(X_JS), len(X_JS)))
for i in range(len(X_JS)):
    for j in range(len(X_JS)):
        arr_js[(i),(j)] = jaccard_binary(X_JS[i], X_JS[j])

In [20]:
df_js = pd.DataFrame(arr_js)
df_js.head()

0         1         2         3     4         5         6     7     \
0  1.000000  0.666667  0.250000  0.250000   0.0  0.000000  0.333333   0.0   
1  0.666667  1.000000  0.333333  0.333333   0.0  0.000000  0.500000   0.0   
2  0.250000  0.333333  1.000000  1.000000   0.0  0.333333  0.500000   0.0   
3  0.250000  0.333333  1.000000  1.000000   0.0  0.333333  0.500000   0.0   
4  0.000000  0.000000  0.000000  0.000000   1.0  0.000000  0.000000   1.0   

   8         9     ...      2195      2196      2197      2198      2199  \
0  0.20  0.000000  ...  0.250000  0.666667  0.250000  0.666667  0.250000   
1  0.00  0.000000  ...  0.333333  1.000000  0.333333  1.000000  0.333333   
2  0.25  0.333333  ...  0.333333  0.333333  0.000000  0.333333  0.333333   
3  0.25  0.333333  ...  0.333333  0.333333  0.000000  0.333333  0.333333   
4  0.25  0.333333  ...  0.333333  0.000000  0.000000  0.000000  0.333333   

       2200      2201      2202      2203      2204  
0  0.250000  0.000000  0.333333  0.000000  0.200000  
1  0.333333  0.000000  0.500000  0.000000  0.000000  
2  0.333333  0.333333  0.500000  0.333333  0.000000  
3  0.333333  0.333333  0.500000  0.333333  0.000000  
4  0.333333  0.333333  0.000000  0.000000  0.666667  

[5 rows x 2205 columns]

## 6. Clustering k-Means Algorithm

In the previous calculations, different ways of computing distances and similarities were presented. One way to make use of that is to cluster the data in certain groups. In the following, the k-Mean Algorithm is used and clusters the dataframe into 5 different groups. The algorithm is hereby based on the euclidean distance. 

In [23]:
from sklearn.cluster import KMeans
kmeans = KMeans(n_clusters=5, random_state=0).fit(df1)
li = kmeans.labels_
li = li.tolist()
df1['Cluster'] = li
df1.head()

Income  Kidhome  Teenhome  Year_Birth  Recency  NumDealsPurchases  \
0  84835.0        0         0        1970        0                  1   
1  57091.0        0         0        1961        0                  1   
2  67267.0        0         1        1958        0                  1   
3  32474.0        1         1        1967        0                  1   
4  21474.0        1         0        1989        0                  2   

   NumCatalogPurchases  NumStorePurchases  NumWebPurchases  NumWebVisitsMonth  \
0                    4                  6                4                  1   
1                    3                  7                7                  5   
2                    2                  5                3                  2   
3                    0                  2                1                  7   
4                    1                  2                3                  7   

   Cluster  
0        4  
1        1  
2        1  
3        2  
4        2

In the dataframe above, the column "Cluster" gives information about which cluster the customer is assigned to. 
To get more insights about the distribution of the customers in the clusters, the following output gives more information. By that, it get's clear that the customers are not equally distributed, especially in the third cluster. 

In [24]:
unique, counts = np.unique(li, return_counts=True)
print(np.asarray((unique, counts)).T)

[[  0 677]
 [  1 672]
 [  2 460]
 [  3   1]
 [  4 406]]


## Source

[1] Jack Daoud; Marketing Analytics; https://www.kaggle.com/jackdaoud/marketing-data

-----------------------------------------------------------
Author: Franz Just